### PR TITLE
Automatic centroid and centroid connector creation

### PR DESCRIPTION
**Pull request by pedrocamargo**
This includes a little more than would be ideal, but that's how I ended up doing it.

- Adds methods to compute network extent and convex hull (convenience methods in reality)
- Adds methods to compute zoning extent and coverage (more convenience methods )
- Adds method for the creation of centroid connectors based on k-means clustering
- Documentation & Testing of the above items
- Code cleaning throughout to avoid the "pass" statement

### CODE REVIEW

File: docs/source/examples/plot_create_zoning.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+# %%
+# We create an empty project on an arbitrary folder
+fldr = join(gettempdir(), uuid4().hex)
+
+# Let's use the Nauru example project for display
+project = create_example(fldr, 'nauru')
+
+# %%
+# We said we wanted 200 zones
```
Comment (janzill) [inline docs/source/examples/plot_create_zoning.py:45]: no you didn't :)

File: aequilibrae/project/zoning.py [hunk 2]
```diff
@@ -35,23 +40,61 @@ class Zoning:
...
+        self.__fields = []
         if self.__has_zoning():
             self.__load()
 
-    def create(self):
+    def new(self, zone_id: int) -> Zone:
+        """Creates a new zone
+
+        Returns:
+            *zone* (:obj:`Zone`): A new link object populated only with link_id (not saved in the model yet)
```
Comment (janzill) [inline aequilibrae/project/zoning.py:56]: copy paste?

File: aequilibrae/project/network/connector_creation.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+from scipy.spatial.distance import cdist
+import shapely.wkb
+from shapely.geometry import LineString
+from aequilibrae.project.database_connection import database_connection
+from aequilibrae.project import network
+from aequilibrae import logger
+
+
+def connector_creation(geo, zone_id: int, srid: int, mode_id: str, link_types='', connectors=1):
+    if len(mode_id) > 1:
```
Comment (janzill) [inline aequilibrae/project/network/connector_creation.py:14]: There's a lot happening here, a docstring would be nice

File: aequilibrae/project/network/connector_creation.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+    for node_to_connect in nds:
+        link = links.new()
+        node_to = proj_nodes.get(node_to_connect)
+        link.geometry = LineString([node.geometry, node_to.geometry])
+        link.modes = mode_id
+        link.direction = 0
+        link.link_type = 'centroid_connector'
+        link.name = f'centroid connector zone {zone_id}'
+        link.capacity_ab = 99999
```
Comment (janzill) [inline aequilibrae/project/network/connector_creation.py:118]: Should we keep the default capacity as a constant in one place for clarity?

File: aequilibrae/project/network/node.py [hunk 5]
```diff
@@ -76,6 +79,7 @@ def renumber(self, new_id: int):
             *new_id* (:obj:`int`): New node_id
         """
 
+        new_id = int(new_id)
         if new_id == self.node_id:
             raise ValueError('This is already the node number')
```
Comment (janzill) [inline aequilibrae/project/network/node.py:84]: Should this maybe be a warning and a do-nothing return?

File: aequilibrae/project/network/node.py [hunk 6]
```diff
@@ -106,14 +111,38 @@ def __save_existing_node(self):
...
         data.append(self.node_id)
         sql = f'Update Nodes set {txts}'
         return data, sql
 
+    def connect_mode(self, area: Polygon, mode_id: str, link_types='', connectors=1):
+        """Adds centroid connectors for the desired mode to the network file
+
+           Centroid connectors are created by clustering all nodes inside the zone that
+           satisfy the mode and link_types criteria in as many clusters as requested connectors.
+           Same algorithm as applied when
```
Comment (janzill) [inline aequilibrae/project/network/node.py:127]: incomplete sentence? Am I missing something?

**Review by janzill — APPROVED** [review]
Just a few minor comments but looks good overall I think

**Comment by janzill** [comment]
Good thing you added automatic cleaning of disconnected nodes yesterday, getting the extend of the model might've gone haywire with a hidden disconnected node somewhere far away. Speaking from experience...

File: docs/source/examples/plot_create_zoning.py (hunk 1 above)
Comment (pedrocamargo) [reply docs/source/examples/plot_create_zoning.py:45]: Good catch!

File: aequilibrae/project/zoning.py (hunk 2 above)
Comment (pedrocamargo) [reply aequilibrae/project/zoning.py:56]: That obvious? :-D

File: aequilibrae/project/network/connector_creation.py (hunk 3 above)
Comment (pedrocamargo) [reply aequilibrae/project/network/connector_creation.py:14]: The docstrings are in the zone and node classes. This guy is not intended to be used independently. Does that make sense?

File: aequilibrae/project/network/connector_creation.py (hunk 4 above)
Comment (pedrocamargo) [reply aequilibrae/project/network/connector_creation.py:118]: Addressed

File: aequilibrae/project/network/node.py (hunk 5 above)
Comment (pedrocamargo) [reply aequilibrae/project/network/node.py:84]: Better indeed.

File: aequilibrae/project/network/node.py (hunk 6 above)
Comment (pedrocamargo) [reply aequilibrae/project/network/node.py:127]: Good catch. corrected.

**Pull request merged**